crest/Assets/Crest/Crest/Shaders/OceanHelpersNew.hlsl

```diff
@@ -32,26 +35,48 @@ float2 IDtoUV(in float2 i_id, in float i_width, in float i_height)
 }
 
 // Sampling functions
-void SampleDisplacements(in Texture2DArray i_dispSampler, in float3 i_uv_slice, in float i_wt, inout float3 io_worldPos, inout half io_sss)
+
+// Displacements. Variance is a statistical measure of how many waves are in the smaller cascades (below this cascade slice). This gives a measure
+// of how much wave content is missing when we sample a particular LOD, and can be used to compensate. The foam sim uses it to compensate for missing
+// waves when computing surface pinch.
+void SampleDisplacements(in Texture2DArray i_dispSampler, in float3 i_uv_slice, in float i_wt, inout float3 io_worldPos, inout half io_variance)
 {
 	const half4 data = i_dispSampler.SampleLevel(LODData_linear_clamp_sampler, i_uv_slice, 0.0);
 	io_worldPos += i_wt * data.xyz;
-	io_sss += i_wt * data.a;
+	io_variance += i_wt * data.w;
+}
+
+void SampleDisplacements( in Texture2DArray i_dispSampler, in float3 i_uv_slice, in float i_wt, inout float3 io_worldPos )
+{
+	half unusedVariance = 0.0;
+	SampleDisplacements( i_dispSampler, i_uv_slice, i_wt, io_worldPos, unusedVariance );
 }
 
 void SampleDisplacementsNormals(in Texture2DArray i_dispSampler, in float3 i_uv_slice, in float i_wt, in float i_invRes, in float i_texelSize, inout float3 io_worldPos, inout half2 io_nxz, inout half io_sss)
 {
 	const half4 data = i_dispSampler.SampleLevel(LODData_linear_clamp_sampler, i_uv_slice, 0.0);
-	io_sss += i_wt * data.a;
 	const half3 disp = data.xyz;
 	io_worldPos += i_wt * disp;
 
-	float3 n; {
-		float3 dd = float3(i_invRes, 0.0, i_texelSize);
-		half3 disp_x = dd.zyy + i_dispSampler.SampleLevel(LODData_linear_clamp_sampler, i_uv_slice + float3(dd.xy, 0.0), dd.y).xyz;
-		half3 disp_z = dd.yyz + i_dispSampler.SampleLevel(LODData_linear_clamp_sampler, i_uv_slice + float3(dd.yx, 0.0), dd.y).xyz;
+	float3 dd = float3(i_invRes, 0.0, i_texelSize);
+	half3 disp_x = dd.zyy + i_dispSampler.SampleLevel( LODData_linear_clamp_sampler, i_uv_slice + float3(dd.xy, 0.0), 0.0 ).xyz;
+	half3 disp_z = dd.yyz + i_dispSampler.SampleLevel( LODData_linear_clamp_sampler, i_uv_slice + float3(dd.yx, 0.0), 0.0 ).xyz;
+
+	// Normal
+	float3 n;
+	{
 		n = normalize(cross(disp_z - disp, disp_x - disp));
 	}
+
+	// SSS - based off pinch
+	{
+		// Don't use variance - yet
+		// float varianceBeforeThisCascade = data.w;
+		float4 du = float4(disp_x.xz, disp_z.xz) - disp.xzxz;
+		float det = (du.x * du.w - du.y * du.z) / (i_texelSize * i_texelSize);
+		io_sss += i_wt * saturate( .03 * (20.0 - (det * 4.0 /*- varianceBeforeThisCascade * 20.0*/)) );
+	}
```

Comment:
This could be wrapped in *_SUBSURFACESCATTERING_ON* to be explicit. But I think this should be stripped from the shader if that is not true anyway since the value would never be used.

Reply:
makes sense, done